flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/StateAssignmentOperation.java

```diff
@@ -201,31 +261,33 @@ private void assignTaskStateToExecutionJobVertices(
 
 		for (int subTaskIndex = 0; subTaskIndex < newParallelism; subTaskIndex++) {
 
-			Execution currentExecutionAttempt = executionJobVertex.getTaskVertices()[subTaskIndex]
-				.getCurrentExecutionAttempt();
+			if (subTaskIndices.contains(subTaskIndex)) {
```

Comment:
instead of `for i in (0 .. newParallelism) -> `contains(i)`, why not supply an `Iterable<Integer> subtaskIDs` instead and then `for (int subtask : subtaskIDs)`? The old codepath would just pass in an iterable from 0 to new parallelism.